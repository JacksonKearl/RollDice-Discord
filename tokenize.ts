export type TokenMatcher =
  | { pattern: string }
  | {
      pattern: string | RegExp
      name: string
    }

interface NormalizedTokenMatcher {
  pattern: RegExp
  name: string
}

export interface TokenMatchResult extends NormalizedTokenMatcher {
  match: string
}

const matchOperators = /[|\\{}()[\]^$+*?.-]/g
const selectInner = /^\/(.*)\/.*$/
const asRegexString = (str: string | RegExp) =>
  typeof str === "string"
    ? str.replace(matchOperators, "\\$&")
    : str.toString().match(selectInner)![1]

export class Tokenizer {
  private matchers: NormalizedTokenMatcher[]
  private matchAll: RegExp

  constructor(matchers: TokenMatcher[]) {
    this.matchers = matchers.map(matcher => ({
      pattern: new RegExp(asRegexString(matcher.pattern)),
      name: "name" in matcher ? matcher.name : matcher.pattern
    }))
    this.matchAll = new RegExp(
      matchers.map(matcher => asRegexString(matcher.pattern)).join("|"),
      "g"
    )

    this.tokenize = this.tokenize.bind(this)
  }

  tokenize(str: string): TokenMatchResult[] {
    let tokens = str.match(this.matchAll)
    if (!tokens) throw new Error("Could not tokenize")
    return tokens.map(match => ({
      ...this.matchers.find(matcher => matcher.pattern.test(match))!,
      match
    }))
  }
}
