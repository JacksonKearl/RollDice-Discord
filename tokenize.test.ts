import { Tokenizer } from "./tokenize"
import * as assert from "assert"

describe("Tokenizer", () => {
  it("can tokenize with simple string matching", () => {
    const { tokenize } = new Tokenizer([
      { pattern: "*" },
      { pattern: "/" },
      { pattern: "+" },
      { pattern: "-" },
      { pattern: "(" },
      { pattern: ")" }
    ])

    const matches = tokenize("  ( + ) / ( * )   -   (  +     )   ")
    expect(matches).toMatchSnapshot()
    expect(matches.length).toBe(11)
    expect(matches[0].match).toBe("(")
    expect(matches[0].name).toBe("(")
  })

  it("can tokenize with regular expression matching", () => {
    const tokenizer = new Tokenizer([
      { pattern: /\/\/.*$/, name: "Comment" },
      { pattern: "+" },
      { pattern: "-" },
      { pattern: /\w+/, name: "Name" }
    ])

    const matches = tokenizer.tokenize("bc + dsf - fasdas // the comment + - dasdas")

    expect(matches).toMatchSnapshot()

    expect(matches.length).toBe(6)
    expect(matches[0].name).toBe("Name")
    expect(matches[0].match).toBe("bc")
    expect(matches[5].name).toBe("Comment")
    expect(matches[5].match).toBe("// the comment + - dasdas")
  })
})
