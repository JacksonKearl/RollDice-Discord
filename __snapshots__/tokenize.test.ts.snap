// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokenizer can tokenize with regular expression matching 1`] = `
Array [
  Object {
    "match": "bc",
    "name": "Name",
    "pattern": /\\\\w\\+/,
  },
  Object {
    "match": "+",
    "name": "+",
    "pattern": /\\\\\\+/,
  },
  Object {
    "match": "dsf",
    "name": "Name",
    "pattern": /\\\\w\\+/,
  },
  Object {
    "match": "-",
    "name": "-",
    "pattern": /\\\\-/,
  },
  Object {
    "match": "fasdas",
    "name": "Name",
    "pattern": /\\\\w\\+/,
  },
  Object {
    "match": "// the comment + - dasdas",
    "name": "Comment",
    "pattern": /\\\\/\\\\/\\.\\*\\$/,
  },
]
`;

exports[`Tokenizer can tokenize with simple string matching 1`] = `
Array [
  Object {
    "match": "(",
    "name": "(",
    "pattern": /\\\\\\(/,
  },
  Object {
    "match": "+",
    "name": "+",
    "pattern": /\\\\\\+/,
  },
  Object {
    "match": ")",
    "name": ")",
    "pattern": /\\\\\\)/,
  },
  Object {
    "match": "/",
    "name": "/",
    "pattern": /\\\\//,
  },
  Object {
    "match": "(",
    "name": "(",
    "pattern": /\\\\\\(/,
  },
  Object {
    "match": "*",
    "name": "*",
    "pattern": /\\\\\\*/,
  },
  Object {
    "match": ")",
    "name": ")",
    "pattern": /\\\\\\)/,
  },
  Object {
    "match": "-",
    "name": "-",
    "pattern": /\\\\-/,
  },
  Object {
    "match": "(",
    "name": "(",
    "pattern": /\\\\\\(/,
  },
  Object {
    "match": "+",
    "name": "+",
    "pattern": /\\\\\\+/,
  },
  Object {
    "match": ")",
    "name": ")",
    "pattern": /\\\\\\)/,
  },
]
`;
